File: PROLOG/act_II/giants_sea.pl
:- multifile describe/1.

describe(giants_sea) :- !,
    write("You have entered the territory of dangerous giants! They begin hurling massive stones at your ships.\n"),
    crew(CurrentCrew),
    Loss is max(20, round(0.2 * CurrentCrew)),
    crew_death(Loss),
    format("\nThe giants' attack reduces your crew by ~d members.\n", [Loss]),
    write("\nYou have no choice but to retreat south.\n"),
    sail(south).

File: HASKELL/act_II/giants_sea.hs
module ActII.GiantsSea where
import Types
import WorldMap

giantsSea :: State -> IO State
giantsSea state = do
    let currentCrew = crew state
    let loss = max 20 (min currentCrew (round (0.2 * fromIntegral currentCrew)))
    let newCrewCount = currentCrew - loss
    let updatedState = state { crew = newCrewCount }

    putStrLn $ "Current crew before loss: " ++ show currentCrew
    putStrLn $ "Calculated loss: " ++ show loss
    putStrLn $ "New crew count after loss: " ++ show newCrewCount
    putStrLn $ "The giants' attack reduces your crew by " ++ show loss ++ " members."
    putStrLn "You have no choice but to retreat south."
    let newplace = circe_sea
    let finalState = updatedState { you_are_at = newplace}
    return finalState